### DISCUSSION TITLE
[Question][Hw2] where is Llvm module?

### DISCUSSION THREAD

**Discussion by ShinWonho** [prosyslab-classroom/cs524-program-analysis#416] [Q&A]
Hi, I want to test some codes in `utop` with `#use` as in HW1.
However, it is impossible to do so, because the Llvm module is unbound.
![image](https://user-images.githubusercontent.com/50018375/133553893-5d47f873-b75a-46ba-8cce-939b49d8aab7.png)

So, my question is,
1. where is the Llvm module?
2. how the Llvm module is binded?
3. can I use utop in a similar way in HW1?

**Comment by kangwoosukeq** [comment]
1. Llvm module is a part of llvm library which handles llvm IR. You can find the interface of module in ```/home/student/.opam/prosyslab-classroom-4.11.0/lib/llvm/llvm.mli``` for both docker and kcloud.

2. The ```src/dune``` file binds the llvm library to your project, so you can use Llvm module in your source code. For more information, check https://dune.readthedocs.io/en/stable/quick-start.html.

3. You can you ```#require``` directive to load the library in ```utop```. For example,
```
# #require "llvm";;
# #use "src/domain.ml";;
```

**Comment by Maokami** [comment]
I think the followings will work:
```
#require "llvm";;
#load "str.cma";;
#require "llvmutils";;
#mod_use "src/domain.ml";;
#use "src/semantics.ml";;
```

Then you can freely use your implemented functions.
```
utop # Value.Int 3;;
- : Value.t = Domain.Value.Int 3
```

Oops, duplicated answer..

**Comment by ShinWonho** [comment]
Thanks a lot! :)
It helps me understand the materials!
